Allow default applications to be localised

jkimbo opened the pull request:
When loading applications in the app launcher look for localised versions of the app as well.

N.B. also added some basic tests for `AppData` using pytest.

## TODO

* [x] Add `en_QQ` versions of the default apps

tombettany commented on kano_apps/AppData.py:
```diff
@@ -105,21 +111,32 @@ def get_applications(parse_cmds=True):
 
-        filtered_apps.append(app)
+    # get localised apps
+    current_locale, _ = locale.getlocale()
```
Probably best to avoid this idiom as it overwrites `_()` in this function which can be confusing if expected later on.

jkimbo replied:
👍

tombettany commented on kano_apps/AppData.py:
```diff
@@ -91,10 +92,15 @@ def get_applications(parse_cmds=True):
+
+    def _collect_apps(application_dir):
+        if not os.path.exists(application_dir):
+            return None
```
Returning `None` is possible but calling a `dict` specific function later on.

jkimbo replied:
Good spot! Returning an empty dict now.

tombettany commented on kano_apps/AppData.py:
```diff
@@ -105,21 +111,32 @@ def get_applications(parse_cmds=True):
+            apps.update(localised_apps)
+
+    filtered_apps = [
+        app for f, app in apps.items()
```
We are using python 2.7 so `.items()` returns a copy of a potentially large list - use `.iteritems()` instead.

jkimbo replied:
👍

tombettany commented on kano_apps/AppData.py:
```diff
@@ -105,21 +111,32 @@ def get_applications(parse_cmds=True):
-                        blacklist += data["overrides"]
+                        blacklist.extend(data["overrides"])
 
             if f[-8:] == ".desktop" and f[0:5] != "auto_":
```
While not explicitly to do with your changes, would be both clearer and more performant if this were an `elif` - the file extension can't be both `.app`  and `.desktop`.

jkimbo replied:
Good point! Fixed

tombettany commented on tests/kano_apps/test_AppData.py:
```diff
@@ -0,0 +1,115 @@
+TEST_APP_ES = {
+    "title": "Pruebas",
+    "tagline": "Aplicación de prueba",
+    "description": "Esta es una aplicación de prueba para, así, las cosas de la prueba.",
```
It might be wise to have a test case (either modify this one or create another) which explicitly uses unicode characters.

tombettany replied:
Actually, this might already have that. If so, ignore the above comment.

jkimbo replied:
Yep this one should test having unicode in the app definition.

tombettany commented:
@jkimbo Looks pretty good, only very minor concerns.

tombettany commented:
@jkimbo You want to merge this or are you going to append the 2 todos to this PR?

jkimbo commented:
@tombettany I'll raise a separate PR for the current TODO's but I am going to add some en_QQ "translations" of the default apps to this PR so that I can test it fully.

jkimbo commented:
Added the `en_QQ` apps and switched to using `kano_settings.system.locale` to get the current locale. Tested on the pi and it all looks good. @tombettany if you're happy with the changes I'll merge and send to QA.

tombettany commented:
@jkimbo Looks fine to me. Feel free to merge.

jkimbo commented:
Thanks @tombettany !
